src/Makefile.am

```diff
@@ -234,6 +234,8 @@ squid_SOURCES = \
 	carp.cc \
 	cbdata.cc \
 	cbdata.h \
+	ClientActiveRequests.h \
+	ClientActiveRequests.cc \
```

Comment:
On the surface, adding two source files just to declare a global variable feels like an overkill. Have you considered adding that global to the existing files instead (e.g., client_side_request.{cc,h})?

Reply:
It us used by both request and reply. I believe having an own file increases clarity and consistency; it used to be in dlink_list.cc :(

Reply:
> I believe having an own file increases clarity and consistency

Yes, but it is still an _overkill_. The global in question is closely tied to the ClientHttpRequest. The global does not deserve its own pair of source code files.

Many folks, myself included, think that even C++ classes do not always deserve their own pair of source code files. The best granularity balances well-known overheads of source code files and well-known overheads caused by separating related declarations with well-known overheads caused by grouping unrelated declarations together. Dedicating two new source code files to a global closely tied to ClientHttpRequest clearly does not achieve the right balance IMO.

Reply:
I'm of two minds here, but I won't go into the philosophical question.
I think we can agree that that symbol doesn't belong with dlink. Unfortunately it's tied to both request and reply, or I would agree with you.
In a pure c++ world we may have request and reply inherit from a common class and do the bookkeeping there RAII-style, but it would still be a couple of additional files

Reply:
> I think we can agree that that symbol doesn't belong with dlink

Yes, of course.

> Unfortunately it's tied to both request and reply

Replies need this global because, AFAICT, somebody wanted the cache manager to list recently active requests first, but that presentation order has nothing to with the proper place for _defining_ this global. I can imagine ten more places in the code where some equivalent of makeThisHead() would be called, all of them outside of client_side_reply.cc. That does not mean ClientActiveRequests is tied to all of those ten places or somehow deserves two dedicated source files just because many places want to reorder cache manager output.

One cannot properly use ClientActiveRequests global without knowing what ClientHttpRequest is, and the ClientHttpRequest object lifetime defines the lifetime of the corresponding ClientActiveRequests member. The global and the type are very strongly tied together.

> In a pure c++ world we may have request and reply inherit from a common class...

FWIW, the whole idea of splitting a single transaction into ClientHttpRequest and clientReplyContext types was probably a mistake in itself, but that is not something we can fix in this PR. Just ignore the fact that clientReplyContext reaches into ClientHttpRequest internals (as if the two were the same object) and place ClientActiveRequests where its objects are defined -- with ClientHttpRequest.

If you want to go a step further, move makeThisHead() into a global TouchClientRequest() function and hide ClientHttpRequest inside client_side_request.cc.

Reply:
> If you want to go a step further, move makeThisHead() into a global TouchClientRequest() function and hide ClientHttpRequest inside client_side_request.cc.

If you don't mind, I'll do it in a followup PR